### PR TITLE
[Merged by Bors] - text aspect ratio bug fix

### LINKED ISSUE #6748: MeasureFunc Text aspect ratio bug
## Bevy version

0.9

& identical behaviour observed with the taffy-2 PR

## What you did

```rust
use bevy::prelude::*;

const MESSAGES: [&str; 8] = [
    "123",
    "12345",
    "1234567",
    "1234567890",
    "123\nxyz",
    "12345\nxyz",
    "1234567\nxyz",
    "1234567890\nxyz",
];

fn main() {
    App::new()
        .add_plugins(DefaultPlugins)
        .add_startup_system(setup)
        .add_system(update)
        .run();
}
fn setup(mut commands: Commands, asset_server: Res<AssetServer>) {
    commands.insert_resource(UiScale { scale: 1.0 });
    commands.spawn(Camera2dBundle::default());
    commands
        .spawn(NodeBundle {
            style: Style {
                align_items: AlignItems::Center,
                justify_content: JustifyContent::Center,
                size: Size::new(Val::Percent(100.), Val::Percent(100.)),                
                ..Default::default()
            },
            ..Default::default()
        })
        .with_children(|builder| {
            builder
                .spawn(NodeBundle {
                    style: Style {
                        size: Size::new(Val::Percent(80.), Val::Percent(80.)),
                        flex_direction: FlexDirection::Column,
                        ..Default::default()
                    },
                    background_color: BackgroundColor(Color::NAVY),
                    ..Default::default()
                })
                .with_children(|builder| {
                    builder.spawn((TextBundle {
                        text: Text::from_section(
                            "".to_string(),
                            TextStyle {
                                font: asset_server.load("fonts/FiraSans-Bold.ttf"),
                                font_size: 24.0,
                                color: Color::WHITE,
                            },
                        ),
                        ..Default::default()
                    },
                    BackgroundColor(Color::RED))
                );
                });
        });
}

pub fn update(
    mut t: Local<f32>,
    mut i: Local<usize>,
    time: Res<Time>,
    mut text_query: Query<&mut Text>,
) {
    *t -= time.delta_seconds();
    if *t <= 0. {
        *t = 1.;
        *i = (*i + 1) % MESSAGES.len();
        let mut text = text_query.single_mut();
        text.sections[0].value = MESSAGES[*i].into();
    }
}
```

## What went wrong

For the single-line text, the length of the cross-axis (which is vertical here) changes depending on the width of the text.
But for multiline text, the length of the cross-axis is constant even though the width of the text changes.

If you change the FlexDirection, the behaviour reverses and the cross-axis varies with the widths of the multiline text and is constant for single-line text.

### PR DESCRIPTION
**Pull request by ickshonpe**
## Objective 

Bevy UI uses a `MeasureFunc` that preserves the aspect ratio of text, not just images. This means that the extent of flex-items containing text may be calculated incorrectly depending on the ratio of the text size compared to the size of its containing node.

Fixes #6748 
Related to #6724

with Bevy 0.9:

Capture_cols_0 9

with this PR (accurately matching the behavior of Flexbox):

Capture_fixed

## Solution
Only perform the aspect ratio calculations if the uinode contains an image.

## Changelog
* Added a field `preserve_aspect_ratio` to `CalculatedSize`
* The `MeasureFunc` only preserves the aspect ratio when `preserve_aspect_ratio` is true.
* `update_image_calculated_size_system` sets `preserve_aspect_ratio` to true for nodes with images.

### CODE REVIEW

File: crates/bevy_ui/src/flex/mod.rs [hunk 1]
```diff
@@ -77,6 +77,7 @@ impl FlexSurface {
         style: &Style,
         calculated_size: CalculatedSize,
         scale_factor: f64,
+        preserve_aspect_ratio: bool,
```
Comment (alice-i-cecile) [inline crates/bevy_ui/src/flex/mod.rs:80]: The logic of when we want to preserve the aspect ratio needs to be clearly documented in the code. Ideally in doc strings on this method.

The PR description is great, but not durable :)

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]

File: crates/bevy_ui/src/flex/mod.rs (hunk 1 above)
Comment (ickshonpe) [reply crates/bevy_ui/src/flex/mod.rs]: Yeah, I wasn't happy with the implementation, querying for images seemed dumb. Rewrote it when I woke up this morning so it adds a preserve aspect ratio flag to CalculatedSize instead, which seems so much better

**Review by alice-i-cecile — APPROVED** [review]
That is a dramatically better solution, nice work.

**Review by Weibye — APPROVED** [review]

**Review by mockersf — APPROVED** [review]

**Comment by james7132** [comment]
bors retry

**Comment by james7132** [comment]
bors retry

**Pull request closed without merging**